File: src/EvalApply.hs
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE ViewPatterns #-}

module EvalApply
  ( eval,
    evalList,
  )
where

import Control.Monad.Trans.Except (except, throwE)
import Data.Text.Format (format)
import Relude hiding (Text)
import qualified System.Exit as Exit
import Text.RawString.QQ (r)
import Types as T
  ( Env,
    Exp (..),
    ScmErr (..),
    fromOuter,
    insertValue,
    isDefined,
    lookup,
    scmNil,
    setValue,
    toConsCell,
  )

handleLambda :: Exp -> [Exp] -> IORef Env -> ExceptT ScmErr IO Exp
handleLambda exp' xs envBox = do
  func <- eval exp' envBox
  args <- xs & mapM (`eval` envBox)
  apply func args

evalList :: [Exp] -> IORef Env -> ExceptT ScmErr IO Exp
evalList xs envBox = case nonEmpty xs of
  Nothing -> return scmNil
  Just xs' -> do
    init xs' & mapM_ (`eval` envBox)
    eval (last xs') envBox

eval :: Exp -> IORef Env -> ExceptT ScmErr IO Exp
-- Self-evaluating types.
eval i@(ScmInt _) _ = return i
eval d@(ScmDouble _) _ = return d
eval str@(ScmStr _) _ = return str
eval (ScmSym s) envBox = case s of
  -- Booleans and other unchangeable constants.
  -- (No, we should not learn Python 2, where the booleans
  -- are part of the prelude!)
  "#t" -> return $ ScmBool True
  "#f" -> return $ ScmBool False
  -- The empty list alias.
  "nil" -> return scmNil
  -- Variable evaluation by name.
  _ ->
    s & (`T.lookup` envBox)
      & maybeToExceptT
        (ScmErr $ format [r|eval: Symbol "{}" undefined.|] [s])
-- Function calls and keywords.
eval (ScmList l) envBox = case l of
  [] -> throwE $ ScmErr "eval: got empty List"
  -- Inline anonymous function invocation.
  -- eg. ((lambda (x) (+ x 2)) 3) ;; => 5
  (lambda@(ScmList _) : xs) -> handleLambda lambda xs envBox
  -- Quote.
  (ScmSym "quote") : xs -> case xs of
    [ScmList l'] -> return $ toConsCell l'
    [quotee] -> return quotee
    _ -> throwE $ ScmErr "quote: nothing to quote"
  (ScmSym "lambda") : xs -> do
    -- ! Here we want to clone a pointer, not to clone an Env.
    -- body := (ScmList(ScmList(vars) : defs))
    ScmClosure (ScmList xs) <$> lift (newIORef (fromOuter envBox))
  -- Definition.
  (ScmSym "define") : xs -> case xs of
    -- Simple definition.
    -- eg. (define f (lambda (x y) *defns*))
    [ScmSym sym, def] -> do
      evalDef <- eval def envBox
      lift $ insertValue sym evalDef envBox
      return scmNil
    -- Syntax sugar for function definition.
    -- eg. (define (f x y) *defns*)
    -- ->  (define f (lambda (x y) *defns*))
    (ScmList (func@(ScmSym _) : args)) : defs ->
      let explicitLambda = ScmList $ [ScmSym "lambda", ScmList args] ++ defs
          desugared = ScmList [ScmSym "define", func, explicitLambda]
       in eval desugared envBox
    _ -> throwE $ ScmErr "define: nothing to define"
  -- Variable reset.
  (ScmSym "set!") : xs -> case xs of
    -- We can only reset a value that is already defined.
    [ScmSym sym, def] -> do
      isDefined' <- lift $ isDefined sym envBox
      if isDefined'
        then do
          evalDef <- eval def envBox
          lift $ setValue sym evalDef envBox
          return scmNil
        else throwE $ ScmErr "set!: can only set a already defined value"
    _ -> throwE $ ScmErr "set!: nothing to set"
  -- Conditional expression.
  (ScmSym "if") : t -> case t of
    [cond, then', else'] -> do
      evalCond <- eval cond envBox
      case evalCond of
        ScmBool True -> eval then' envBox
        ScmBool False -> eval else' envBox
        _ -> throwE $ ScmErr "if: expected Boolean"
    _ -> throwE $ ScmErr "if: ill-formed"
  (ScmSym "cond") : t ->
    let evalTail [ScmList [ScmSym "else", then']] = eval then' envBox
        evalTail ((ScmList [cond, then']) : xs) = do
          evalCond <- eval cond envBox
          case evalCond of
            ScmBool True -> eval then' envBox
            ScmBool False -> evalTail xs
            _ -> throwE $ ScmErr "cond: expected Boolean"
        evalTail _ = throwE $ ScmErr "cond: ill-formed"
     in evalTail t
  -- Begin, exit and display expressions.
  (ScmSym "begin") : t -> evalList t envBox
  (ScmSym "exit") : t ->
    let exit [] = lift Exit.exitSuccess
        exit [ScmInt 0] = lift Exit.exitSuccess
        exit [ScmInt x] | x > 0 = lift . Exit.exitWith . Exit.ExitFailure . fromIntegral $ x
        exit [ScmDouble d] = exit [ScmInt $ truncate d]
        exit _ = throwE $ ScmErr "exit: invalid exit code"
     in exit t
  (ScmSym "display") : t ->
    let printElem x = do
          val <- eval x envBox
          lift $ print val
          return $ ScmList []
     in case t of
          [] -> throwE $ ScmErr "display: nothing to display"
          xs -> xs & mapM printElem >> return scmNil
  (ScmSym "newline") : t -> case t of
    [] -> lift $ putStrLn "" >> return scmNil
    _ -> throwE $ ScmErr "newline: expected no arguments"
  -- Function call by name.
  func@(ScmSym _) : t -> handleLambda func t envBox
  _ -> throwE $ ScmErr "eval: unexpected compound Exp"
eval _ _ = throwE $ ScmErr "eval: unexpected Exp"

apply :: Exp -> [Exp] -> ExceptT ScmErr IO Exp
-- func can only be Primitive or Closure
apply (ScmPrimitive prim) args = except $ prim args
apply (ScmClosure body' envBox) args = do
  let (ScmList (ScmList vars : defs)) = body'
  env' <- readIORef envBox & lift
  localEnv <- newIORef env' & lift
  zip vars args
    & mapM_ (\(ScmSym i, arg) -> insertValue i arg localEnv)
    & lift
  evalList defs localEnv
apply _ _ = throwE $ ScmErr "apply: unexpected Exp"

File: src/main/scala/EvalApply.scala
package io.github.rami3l.yascm

import scala.util.{Try,Success,Failure}

extension (env: Env) {
  def handleLambda(func: Exp, args: List[Exp]): Try[Exp] = Try {
    val func1 = env.eval(func).get
    val args1 = args.map { env.eval(_).get }
    apply(func1, args1).get
  }

  def evalList(exps: List[Exp]): Try[Exp] = Try {
    exps.init.foreach { env.eval(_).get }
    env.eval(exps.last).get
  }

  def eval(exp: Exp): Try[Exp] = Try {
    exp match {
      // * Self-evaluating types.
      case n as Num(_) => n
      case s as Str(_) => s

      // * Variable evaluation by name.
      case Sym(s) =>
        env
          .lookup(s)
          .getOrElse(throw Exception(s"eval: Symbol `$s` undefined"))

      // * Function calls and keywords.
      case ScmList(Nil) => throw Exception("eval: got empty function call")
      // Inline anonymous function invocation.
      // eg. ((lambda (x) (+ x 2)) 3) ;; => 5
      case ScmList((func as ScmList(_)) :: xs) => env.handleLambda(func, xs).get
      // Quote.
      case ScmList(Sym("quote") :: xs) =>
        xs match {
          case quotee :: Nil => quotee
          case _             => throw Exception("quote: nothing to quote")
        }
      // Anonymous function literal.
      // eg. (lambda (x y) *defns*)
      case ScmList(Sym("lambda") :: xs) => {
        // ! Here we want to clone a pointer, not to clone an Env.
        val closEnv = Env(outer = env)
        Closure(body = ScmList(xs), closEnv)
      }
      // Definition.
      case ScmList(Sym("define") :: xs) =>
        xs match {
          // Simple definition.
          // eg. (define f (lambda (x y) *defns*))
          case Sym(sym) :: defn :: Nil => {
            val defn1 = env.eval(defn).get
            env.insertVal(sym, defn1)
            ScmNil
          }
          // Syntax sugar for function definition.
          // eg. (define (f x y) *defns*)
          // ->  (define f (lambda (x y) *defns*))
          case ScmList((func as Sym(_)) :: args) :: defns =>
            env.eval {
              ScmList(
                Sym("define")
                  :: func
                  :: ScmList(Sym("lambda") :: ScmList(args) :: defns)
                  :: Nil
              )
            }.get
          case _ => throw Exception("define: nothing to define")
        }
      // Variable reset.
      case ScmList(Sym("set!") :: xs) =>
        xs match {
          // We can only reset a value that is already defined.
          case Sym(sym) :: defn :: Nil if env.lookup(sym).isDefined => {
            val defn1 = env.eval(defn).get
            env.setVal(sym, defn1)
            ScmNil
          }
          case _ => throw Exception("set!: nothing to set")
        }
      // Conditional expression.
      case ScmList(Sym("if") :: cond :: then1 :: else1 :: Nil) => {
        env.eval(cond).get match {
          case Bool(true)  => env.eval(then1).get
          case Bool(false) => env.eval(else1).get
          case _           => throw Exception("if: expected Bool")
        }
      }
      case ScmList(Sym("if") :: _) => throw Exception("if: ill-formed")
      case ScmList(Sym("cond") :: tail) => {
        def evalTail(xs: List[Exp]): Try[Exp] = Try {
          xs match {
            case ScmList(Sym("else") :: then1 :: Nil) :: Nil =>
              env.eval(then1).get
            case ScmList(cond :: then1 :: Nil) :: xs1 =>
              env.eval(cond).get match {
                case Bool(true)  => env.eval(then1).get
                case Bool(false) => evalTail(xs1).get
                case _           => throw Exception("cond: expected Bool")
              }
            case _ => throw Exception("cond: ill-formed")
          }
        }
        evalTail(tail).get
      }
      case ScmList(Sym("display") :: xs) => {
        xs.foreach { println(_) }
        ScmNil
      }
      // Call function by name.
      case ScmList((func as Sym(_)) :: args) => handleLambda(func, args).get

      case _ => throw Exception("eval: unexpected expression")
    }
  }
}

def apply(func: Exp, args: List[Exp]): Try[Exp] = Try {
  func match {
    // `func` can only be Primitive or Closure.
    case Primitive(prim) => prim(args).get
    case Closure(body, env) => {
      val ScmList(ScmList(vars) :: defns) = body
      var localEnv = Env(outer = env)
      vars.zip(args).map { (ident, arg) =>
        localEnv.insertVal(ident.asInstanceOf[Sym].value, arg)
      }
      localEnv.evalList(defns).get
    }
    case _ => throw Exception("apply: unexpected expression")
  }
}
